### PR TITLE
Ensuring necessary shared libraries are included in the conda package

### PR DESCRIPTION
**Pull request by douglowe**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

The current conda package does not include the `umfile.so` file (see this [bug report](https://github.com/conda-forge/cf-python-feedstock/issues/24)). I believe that this is not included because the current build script relies on pip to build all the supporting packages for cf-python, so conda-build then leaves out the shared library because it is referencing files that are not managed by the conda build system - even though these will be provided later by the run requirements, that is not made explicitly clear to conda build.

So, I've added all the required packages to the build section of the script. Doing this enables pip to create the required libraries in my local conda environments, without pip having to install anything but cf-python, so hopefully it will do the same here too.

I've removed the version requirements from the run section - hopefully these will now be automatically set by conda build using the version information from the build section. That might not happen - if it doesn't we should add these back in.

### CODE REVIEW

**Comment by douglowe** [comment]
@conda-forgeadmin, please rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -19,17 +19,26 @@ requirements:
...
     - make
+    - cfdm
+    - cftime
+    - cfunits
+    - netcdf4
+    - netcdf-flattener
+    - numpy
+    - psutil
+    - packaging
+    - pyparsing
```
Comment (ocefpaf) [inline recipe/meta.yaml:31]: Unless these are needed at build time we do not need to specify those here.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -19,17 +19,26 @@ requirements:
...
     - python
     - pip
   run:
     - python
-    - cfdm >=1.9.0.0,<1.9.1.0
-    - cftime >=1.5.0
-    - cfunits >=3.3.3
-    - netcdf4 >=1.5.4
-    - numpy >=1.15
-    - psutil >=0.6.0
```
Comment (ocefpaf) [inline recipe/meta.yaml:32]: This library is heavily pinned upstream: https://github.com/NCAS-CMS/cf-python/blob/v3.13.0/requirements.txt

File: recipe/meta.yaml [hunk 3]
```diff
@@ -11,7 +12,7 @@ source:
   sha256: 8d7f82a9563c0ae46c2725faeca43ee3a2db12a9b75c5a819bcb1b5ef107f322
 
 build:
-  number: 0
+  number: {{ build }}
```
Comment (ocefpaf) [inline recipe/meta.yaml:15]: Please revert this change. It makes the recipe hard to read and this variable is used only once, so no need for it.

**Comment by ocefpaf** [comment]
Can you add a test for this functionality. It seems that it fails with a warning only, so we don't really know if it is working from a green CI. BTW, it is still failing b/c the subprocess call to make is failing to find a compiler:

```
2022-08-08T10:57:57.0678466Z   ********************************************************************************
2022-08-08T10:57:57.0684501Z   Running: make -C /home/conda/feedstock_root/build_artifacts/cf-python_1659956147498/work/build/lib/cf/umread_lib/c-lib
2022-08-08T10:57:57.0689560Z 
2022-08-08T10:57:57.0695950Z   make: Entering directory '/home/conda/feedstock_root/build_artifacts/cf-python_1659956147498/work/build/lib/cf/umread_lib/c-lib'
2022-08-08T10:57:57.0701577Z   gcc -Wall -fPIC -DLINUX -c umfile.c
2022-08-08T10:57:57.0707151Z   make: gcc: No such file or directory
2022-08-08T10:57:57.0712023Z   make: *** [Makefile:47: umfile.o] Error 127
2022-08-08T10:57:57.0718204Z   make: Leaving directory '/home/conda/feedstock_root/build_artifacts/cf-python_1659956147498/work/build/lib/cf/umread_lib/c-lib'
2022-08-08T10:57:57.0723017Z 
2022-08-08T10:57:57.0728811Z    --------------------------------------------------------------------------------
2022-08-08T10:57:57.0733990Z   WARNING: Failed to build the UM read C library.
2022-08-08T10:57:57.0739745Z            Attempting to read UKMO PP and UM format files will result in failure.
2022-08-08T10:57:57.0745119Z            This will not affect any other cf functionality.
2022-08-08T10:57:57.0750483Z            In particular, netCDF file processing is unaffected.
2022-08-08T10:57:57.0755803Z   --------------------------------------------------------------------------------
2022-08-08T10:57:57.0760811Z 
2022-08-08T10:57:57.0765897Z    ********************************************************************************
2022-08-08T10:57:57.0771646Z 
2022-08-08T10:57:57.0776632Z   cf build successful
```

We will need to patch it to find the gcc binary correctly.

File: recipe/meta.yaml (hunk 1 above)
Comment (douglowe) [reply recipe/meta.yaml:31]: When installing cf-python using pip these packages are installed - so they are all needed at build time (even if not directly needed by cf-python).

File: recipe/meta.yaml (hunk 3 above)
Comment (douglowe) [reply recipe/meta.yaml:15]: This is standard practice for conda-forge build recipes.

File: recipe/meta.yaml (hunk 1 above)
Comment (ocefpaf) [reply recipe/meta.yaml:31]: When installing with pip you are at the  "run" phase and then yes, they are needed. But not at build time. Note that none of those are listed as setup requires in the package upstream.

File: recipe/meta.yaml (hunk 3 above)
Comment (ocefpaf) [reply recipe/meta.yaml:15]: Nope. We abandoned that a long time ago. Can you please revert?

File: recipe/meta.yaml (hunk 2 above)
Comment (douglowe) [reply recipe/meta.yaml:32]: I've had numerous problems with conda forge packages which are heavily pinned in the run requirements, but when versions are updated these pins are not updated, and the released conda package ends up either uninstallable, or unuseable, because of those pins. If we can avoid having to use them, I think it would be better for package maintainability.

**Comment by douglowe** [comment]
That's an annoying error (seeing as this is what I'm trying to add). I'll see if I can find a fix.

File: recipe/meta.yaml (hunk 2 above)
Comment (ocefpaf) [reply recipe/meta.yaml:32]: Please open an issue upstream about that. I do update packages with grayskull only prevent stale pins, so you are seeing issues mostly where automerge is used without grayskull.

A lose pin like that without full tests is prone to break and not future proof. Upstream has tons of tests to ensure these intervals work as expected.

**Comment by douglowe** [comment]
@conda-forge-admin, please rerender

**Comment by ocefpaf** [comment]
> That's an annoying error (seeing as this is what I'm trying to add). I'll see if I can find a fix.

The build won't find `gcc` as I mentioned above. You'll need to patch the build to use `$GCC` instead.

**Comment by douglowe** [comment]
@conda-forge-admin, please rerender

**Comment by douglowe** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml (hunk 3 above)
Comment (douglowe) [reply recipe/meta.yaml]: Sure thing - I've reverted this change.

File: recipe/meta.yaml (hunk 2 above)
Comment (douglowe) [reply recipe/meta.yaml]: Ahh, I wasn't aware of grayskull - that tool does look very useful for pip based packages. I'll make a note and suggest it's use next time I have this problem.

I've reverted my changes here.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -36,6 +38,7 @@ test:
     - cf
   commands:
     - pip check
+    - python -c "import unittest; from cf.test import test_pp; unittest.main(test_pp)"
```
Comment (ocefpaf) [inline recipe/meta.yaml:41]: I'm OK running all the tests. We avoided that in the past due to CI limitation but since we moved to Azure that is no longe true and running all tests is probably better. (No need to add it in this PR though.)

File: recipe/meta.yaml (hunk 1 above)
Comment (douglowe) [reply recipe/meta.yaml]: Thanks for the explanation. I've mostly worked with none-pip packages so far, so am not so familiar with the pip build system. I've reverted my changes here too - to see if only the makefile patch is needed to fix this issue.

File: conda-forge.yml [hunk 5]
```diff
@@ -4,3 +4,6 @@ bot:
 github:
   branch_name: main
   tooling_branch_name: main
+azure:
+  store_build_artifacts: True
```
Comment (ocefpaf) [inline conda-forge.yml:8]: You can remove these two. BTW, when on  linux, you can easily test the build locally using our build script: `python build-locally.py 2>&1 | tee log`

File: recipe/meta.yaml (hunk 4 above)
Comment (douglowe) [reply recipe/meta.yaml:41]: I agree on running as many tests as possible. But I found that quite a few of their tests fail when I tried it locally - so for the moment probably we keep to only this test. I'll raise a Q with the original package developers about this though.

**Comment by douglowe** [comment]
@conda-forge-admin, please rerender

File: conda-forge.yml (hunk 5 above)
Comment (douglowe) [reply conda-forge.yml:8]: Yep - I've removed this again now.

File: recipe/meta.yaml (hunk 2 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: We have an experimental functionality to update automatically with grayskull using the bot. When that is out most of the headaches you mentioned should be fixed (we hope).

File: recipe/meta.yaml (hunk 1 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: There are some pathological cases where the runtime dependencies are needed at build time when the package imports itself. That is a bad practice but quite common in old packages. Besides that, you only need the setup requires dependencies at build time (and the non-python ones that are hard to specify using python metadata.)

**Comment by douglowe** [comment]
@conda-forge-admin, please rerender
